tests/contract.rs

```diff
@@ -230,6 +230,169 @@ fn mint_burn() {
     assert_eq!(0, storage_balance);
 }
 
+#[test]
+fn call_increases_contract_asset_balance_and_balance_register() {
+    let rng = &mut StdRng::seed_from_u64(2322u64);
+
+    let gas_limit = 1_000_000;
+    let asset_id: AssetId = rng.gen();
+    let call_amount = 500u64;
+
+    let mut test_context = TestBuilder::new(2322u64);
+    let contract_id = test_context
+        .setup_contract(vec![Opcode::RET(REG_BAL)], None, None)
+        .contract_id;
+
+    let (script_ops, offset) = script_with_data_offset!(
+        data_offset,
+        vec![
+            // load call data to 0x10
+            Opcode::ADDI(0x10, REG_ZERO, data_offset + 32),
+            // load gas forward to 0x11
+            Opcode::ADDI(0x11, REG_ZERO, gas_limit as Immediate12),
+            // load balance to forward to 0x12
+            Opcode::ADDI(0x12, REG_ZERO, call_amount as Immediate12),
+            // load the asset id to use to 0x13
+            Opcode::ADDI(0x13, REG_ZERO, data_offset),
+            // call the transfer contract
+            Opcode::CALL(0x10, 0x12, 0x13, 0x11),
+            Opcode::RET(REG_ONE),
+        ]
+    );
+    let script_data: Vec<u8> = [
+        asset_id.as_ref(),
+        Call::new(contract_id, 0, offset as Word).to_bytes().as_slice(),
+    ]
+    .into_iter()
+    .flatten()
+    .copied()
+    .collect();
+
+    // starting contract balance
+    let start_balance = test_context.get_contract_balance(&contract_id, &asset_id);
+    assert_eq!(start_balance, 0);
+
+    // call contract with some amount of coins to forward
+    let transfer_tx = test_context
+        .gas_limit(gas_limit)
+        .gas_price(0)
+        .byte_price(0)
+        .coin_input(asset_id, call_amount)
+        .contract_input(contract_id)
+        .contract_output(&contract_id)
+        .change_output(asset_id)
+        .script(script_ops)
+        .script_data(script_data)
+        .execute();
+
+    // Ensure transfer tx processed correctly
+    assert!(!transfer_tx.should_revert());
+
+    // verify balance transfer occurred
+    let end_balance = test_context.get_contract_balance(&contract_id, &asset_id);
+    assert_eq!(end_balance, call_amount);
+
+    // verify balance register was set
+    assert_eq!(transfer_tx.receipts()[1].val().unwrap(), call_amount);
```

Comment:
The logic seems correct, but its always safer to just `LOG` the register and then inspect the receipt. The `CALL` receipt is built with the arguments of the call and by itself isn't a guarantee that the register was properly set.